programs/castle-vault/src/instructions/rebalance.rs

```diff
@@ -48,32 +48,51 @@ pub struct Rebalance<'info> {
     #[account(
         mut,
         constraint = !vault.value.last_update.is_stale(clock.slot)? @ ErrorCode::VaultIsNotRefreshed,
-        has_one = solend_reserve,
-        has_one = port_reserve,
-        has_one = jet_reserve,
     )]
     pub vault: Box<Account<'info, Vault>>,
 
-    pub solend_reserve: Box<Account<'info, SolendReserve>>,
-
-    pub port_reserve: Box<Account<'info, PortReserve>>,
-
-    pub jet_reserve: AccountLoader<'info, jet::state::Reserve>,
+    pub solend_reserve: AccountInfo<'info>,
+    pub port_reserve: AccountInfo<'info>,
+    pub jet_reserve: AccountInfo<'info>,
 
     pub clock: Sysvar<'info, Clock>,
 }
 
+type SolendAccount<'info> = Account<'info, SolendReserve>;
+type PortAccount<'info> = Account<'info, PortReserve>;
+type JetAccount<'info> = AccountLoader<'info, jet::state::Reserve>;
+
 impl TryFrom<&Rebalance<'_>> for AssetContainer<Reserves> {
     type Error = ProgramError;
     fn try_from(r: &Rebalance<'_>) -> Result<AssetContainer<Reserves>, Self::Error> {
+        let flags = r.vault.get_yield_source_flags();
+
         // NOTE: I tried pretty hard to get rid of these clones and only use the references.
         // The problem is that these references originate from a deref() (or as_ref())
         // and end up sharing lifetimes with the Context<Rebalance>.accounts lifetime,
         // which means that the lifetimes are shared, preventing any other borrows
         // (in particular the mutable borrow required at the end to save state)
-        let solend = Some(Reserves::Solend(r.solend_reserve.deref().deref().clone()));
-        let port = Some(Reserves::Port(r.port_reserve.deref().deref().clone()));
-        let jet = Some(Reserves::Jet(Box::new(*r.jet_reserve.load()?)));
+        let solend = match flags.contains(YieldSourceFlags::SOLEND) {
```

Comment:
Since we explicitly call `try_from`, the `has_one` is not needed. That's my understanding. Is it correct?

Reply:
1. I've added an explicit pubkey check, good catch.
2. The approach you suggested would trigger a stack access violation error, which I'm still trying to debug.
3. The approach you suggested feels less readable to me compared to the `match`, which is quite compact as well and logically more straightforward.

Reply:
`has_one` checks the key, not the form of the struct. We're replacing it here with the pubkey equality check.

There should be an error if the pubkey fails the equality check since that results in an incorrect rebalance being performed if one or more accounts are wrong. It should fail entirely so as to avoid this.

Reply:
I updated it for Solend and Jet. However it keeps reporting stack access violation for Port. I'm not sure where to use the Box? I tried something like this but didn't work
```
let port = flags
            .contains(YieldSourceFlags::PORT)
            .as_option()
            .map(|()| {
                r.port_reserve.key.eq(&r.vault.port_reserve).as_result(
                    *Box::new( Account::<PortReserve>::try_from(&r.port_reserve) ),
                    ErrorCode::InvalidAccount,
                )?
            })
            .transpose()?
            .map(|a| Reserves::Port(a.deref().clone()));
```
However I wonder why this is happening with this approach, but not with the `match` approach. There is probably one additional copy somewhere what I don't see, that can probably be eliminated?